java/src/main/java/ai/onnxruntime/NativeObject.java

```diff
@@ -0,0 +1,211 @@
+package ai.onnxruntime;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Logger;
+
+/**
+ * This is the base class for anything backed by JNI. It manages open versus closed state and usage
+ * handling.
+ */
+abstract class NativeObject implements AutoCloseable {
+
+  static {
+    try {
+      OnnxRuntime.init();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to load onnx-runtime library", e);
+    }
+  }
+
+  private static final Logger logger = Logger.getLogger(NativeObject.class.getName());
+
+  private final Object handleLock = new Object();
+
+  private final long handle;
+
+  private volatile boolean closed;
+
+  /** This is used to ensure the close operation will only proceed when there are no more usages. */
+  private final AtomicInteger activeUsagesCount;
+
+  NativeObject(long handle) {
+    this.handle = handle;
+    this.closed = false;
+    this.activeUsagesCount = new AtomicInteger(1);
+  }
+
+  /**
+   * Generates a description with the backing native object's handle.
+   *
+   * @return the description
+   */
+  @Override
+  public String toString() {
+    return getClass().getSimpleName() + "@" + Long.toHexString(handle);
+  }
+
+  /**
+   * Check if the resource is closed.
+   *
+   * @return true if closed
+   */
+  public final boolean isClosed() {
+    return closed;
+  }
+
+  /**
+   * This internal method allows implementations to specify the manner in which this object's
+   * backing native object(s) are released/freed/closed.
+   *
+   * @param handle a long representation of the address of the backing native object.
+   */
+  abstract void doClose(long handle);
+
+  /**
+   * Releases any native resources related to this object.
+   *
+   * <p>This method must be called or else the application will leak off-heap memory. It is best
+   * practice to use a try-with-resources which will ensure this method always be called. This
+   * method will block until any active usages are complete.
+   */
+  @Override
+  public void close() {
+    synchronized (handleLock) {
+      if (closed) {
+        return;
+      }
```

Comment:
This should throw some exception that would clearly tell that things are going wrong. Somehow, a thread is operating on an object that was already be destroyed and but it thinks it has a valid handle on it. With property implemented refcounting it should never happen, as it looks like the a thread is holding a reference to it but the object is destroyed anyway. This condition should prompt an assert or an non-recoverable exception. Which is what Segfault is currently doing for you.

Refcounting or not, each native object should have a proper and clear ownership and judjing but this piece of code, you do not expect that refcounting would actually save you. Why introducing it?
You can wrap a session object into your recounted class and save other users from the bearing the expense of it.

Reply:
The close() method is called by developers. It is within reason that somebody may close the object more that once and we do not really care since the object was closed already. An exception is thrown if the session is used again (like if they use run() again).

The active usage (NOT reference) counting is beside the point here, since that is done in private implementation hidden from the users and with try-with-resources.

Reply:
close() is usually the last thing that is being called on the object. Since the calling code seems to be unware that the underlying native object is no longer there, it can call all sorts of functions on it before close(). By this reasoning, one would need to check of the object is still around in every interface which is not being done. So it would still lead you to a segfault, would it not? Discovering the the object is gone at the time of close() does nothing to save you if the program continues to be invalid.  If there is not a clear ownership and the recounting is not implemented right such checks actually do nothing. If two threads are trying to close the object but the recount is 1, both would pass the check, and they would happily destroy the object twice thereby corrupting memory. And you already know where it leads -SEGFAULT.
Let me summarize: this PR introduces atomics and locks, both of which impose additional usage requirements and performance costs on other users for the sake of your usage scenario. However, I feel, that the additional overhead and complexity should be shifted to your code.
On another note, there are inter dependencies on some objects. You will see in the next release, OrtAllocators API being exposed and IOBidning. Those depend on Session. Session depends on Environment. Refcounting them individually would not make sense.

---
In reply to: [471666137](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471666137) [](ancestors = 471666137)

Reply:
The semantics of Java's AutoCloseable interface say:
```Note that unlike the close method of Closeable, this close method is not required to be idempotent. In other words, calling this close method more than once may have some visible side effect, unlike Closeable.close which is required to have no effect if called more than once. However, implementers of this interface are strongly encouraged to make their close methods idempotent.```
So having it accept multiple calls to close is idiomatic.

Reply:
However the upcoming changes to allocators and IObinding do make me worried we'll need a bigger solution to this. The allocators were originally exposed and had a more complicated tracking mechanism, but during the initial integration of the Java API we decided that it wasn't necessary and lots of complexity was removed by making the allocators an internal detail.

Reply:
Thanks for the correction. The underlying Release*() methods  are not idempotent. Nor, I feel, they are correctly made such by this PR. Perhaps, Closable is a better interface to implement.

---
In reply to: [471682951](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471682951) [](ancestors = 471682951)

Reply:
Yes, the allocators we are exposing actually hang of the session. They are dependent on the build and the execution providers present so one can allocate a native piece of memory and bind it to input/output. 
So, facing this user scenario, I would probably refcount the session object so I can swap it and make threads that are just starting they work work with the new session while others are still finishing up with the old one. But everything else, actually depends on the active usage and they way code is organized, contrary to the assumption made in this PR.
In C# we use a trick to add everrything IDisposable to a list so a single object is disposed with everything else inside. So each thread, while sharing a session, can create/destroy other native objects within a single run.
Again, this is a usage scenario. Keeping the API simple allows to face future with confidence.

---
In reply to: [471687336](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471687336) [](ancestors = 471687336)

Reply:
In Java `AutoCloseable` recommends idempotence, `Closeable` requires it. `Closeable` is explicitly documented that calling it twice has no effect.

Reply:
The increment/decrement and synchronization mechanisms would start to show degraded performance in cases where there is great thread contention, which is pretty expected. There is no free lunch when it comes to concurrency as always. With regards to the amount of CPU time this introduces for the simple single-threaded vanilla implementation like in the sample code, since that is a single thread, there should be very very minimal performance hit. The AtomicInteger updates for the active usage counts should be super quick since those are two CAS operations. Since there is only one thread, the CAS is mathematically guaranteed to be successful on the first attempt since no other thread would have attempted to update the count. The synchronized blocks are only relevant on close in the thread doing the close() and the last currently active run which notifies the waiting close thread. It is not a common use case to call close() from many threads at once.

I think that such small investments in processing are worthwhile to ensure safety. To reiterate, the TF project does a similar thing, so if they do it, I think there is enough of a community use case to warrant adding such safety mechanisms.

Reply:
Is there a description of the allocator semantics in the coming updates? I couldn't find a good one when I initially wrote the API, it seemed to be an undocumented facet of the C++ implementation and my C++ isn't good enough to figure out the behaviour from that. Plus undocumented implementation details are very different from documented specs.

Ditto for IOBinding.

Reply:
The documentation gap is recognized internally and there will be an effort to close it. Please, do ask me questions about the behavior.
Please, refer to this PR for more details. 
It also demonstrates the way the API is exposed in C#. [https://github.com/microsoft/onnxruntime/pull/4646](https://github.com/microsoft/onnxruntime/pull/4646)

---
In reply to: [471708975](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471708975) [](ancestors = 471708975)

Reply:
The initial allocator exposed was a default CPU allocator. It was a singleton that never required destruction.
The allocators obtained with the new API CreateAllocator() are wrapping instances that require ReleaeAllocator() call when no longer needed. 
Those refer to the internal Session allocators. 
The internal sessions allocators are tied to the execution providers.  Those depend on two factors: 1) whether it is present in the binary 2) whether it was added in the session options.
For example, if you have a build with CPU and CUDA, you specify then in the SessionOptions, create a Session.
Then you create an instance of OrtMemoryInfo that has all the characteristics of the allocator required. (Note, this is also a native Closable/Disposable class)
Using that you call CreateAllocator() which internally checks if the requested allocator is present within the session. It would then create a wrapper and return it.
You now can allocate memory on the device which returns OrtMemoryAllocation instance.
 This is useful in some scenarios 1) You do not want to reallocate memory for input/output and just re-use a  pre-allocated chunk progressively specifying bigger shapes but re-use the same chunk
2) You want to feed output back to input. So you just re-bind input/output trading places.
3) You have a dynamically shaped output. In this case you can not pre-allocated, but by using OrtMemoryInfo you can specify where to allocate and it will do so.
This are especially useful, if you need it to be on the device or you want to feed input that is already allocated, say on CUDA. So no copying to host memory is needed.
in cases 1 and 2 the caller owns the memory, in case of 3, Onnxruntime owns that (Returned OrtValue does).

---
In reply to: [471751379](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r471751379) [](ancestors = 471751379,471708975)

Reply:
> 3) You have a dynamically shaped output. In this case you can not pre-allocate, but by using OrtMemoryInfo you can specify where to allocate and it will do so.
> This is especially useful, if you need it to be on the device or you want to feed input that is already allocated, say on CUDA. So no copying to host memory is needed.
> in cases 1 and 2 the caller owns the memory, in case of 3, Onnxruntime owns that (Returned OrtValue does).

@Craigacp Here you go, you should consider your Java API not supporting CUDA memory as a "performance problem". No need to run benchmarks to know that copying data around unnecessarily is going to be slow. And, no, the official Java API of OpenCV doesn't support CUDA, so even if you figure out something for ORT, it's still going to be a problem! Why do you refuse the recognize the importance of this? I'm not saying that JavaCPP is perfect, but at least I'm trying to do something about it, while you are not.

Reply:
1) Yes, the allocators are thread-safe. They are the same ones used by the session object and the session object support concurrent execution my multiple threads.
2) Yes, it is possible to bind inputs/outputs to a slub of memory allocated by someone else. When we bind a pre-allocated piece of memory, it is not owned by ORT or the corresponding OrtValue. One of the python binding tests makes use of CUDA memory allocated by PyTorch. Current C# API requires passing OrtMemoryAllocation wrapping the chunk. That currently require an allocator ptr. So I can see that in C# we could make it more convinient to supply 3rd party allocations.

---
In reply to: [474764836](https://github.com/microsoft/onnxruntime/pull/4329#discussion_r474764836) [](ancestors = 474764836)